### ISSUE TITLE
CSRF vulnerability in web-based Interface

### ISSUE THREAD

**Issue by pehelwan** [DanWin/hosting#18]
Hey,

Just letting you know that this has site wide no-CSRF protection .
For example on this endpoint -

https://hosting.danwin1210.me/files.php ,no CSRF tokens are assigned or verified when a user does an action.

Regards,
(Security Researcher)

**Comment by pehelwan** [comment]
CVE reference updated to CVE-2018-7308

**Comment by NoahvdAa** [comment]
This issue is really important as any site can upload/delete/create files if a user is logged in.
For example, if you visit my example page: [https://file.webwevers.nl/c1123d5439d61c7a7821367bdf6da93f.html](https://file.webwevers.nl/c1123d5439d61c7a7821367bdf6da93f.html), and happen to be logged in on the file manager, it will create the file **CSRFShouldBeFixed** in **/www/**. This issue can be fixed very easily however, by just creating a CSRF token on login / on page load, and then storing it in the session:
```php
<?php
$token = sha1(uniqid());
$_SESSION['token'] = $token;
?>
```
Then, in your form use this:
```html
<input type="hidden" name="token" value="<?= $_SESSION['token'] ?>">
```
When the form is submitted, just compare the two tokens.
```php
if($_POST['token'] != $_SESSION['token']){
  echo 'Your message here';
  exit;
}else{
  echo 'Tokens matched, everything is fine.';
  exit;
}
```

**Comment by DanWin** [comment]
Fully fixed in 9c5294e64
